src/main/java/org/broadinstitute/hellbender/utils/downsampling/PositionalDownsampler.java

```diff
@@ -71,16 +73,31 @@ private void handlePositionalChange( final GATKRead newRead ) {
         // Use ReadCoordinateComparator to determine whether we've moved to a new start position.
         // ReadCoordinateComparator will correctly distinguish between purely unmapped reads and unmapped reads that
         // are assigned a nominal position.
-        if ( previousRead != null && ReadCoordinateComparator.compareCoordinates(previousRead, newRead, header) != 0 ) {
-            if ( reservoir.hasFinalizedItems() ) {
-                finalizeReservoir();
+        if ( previousRead != null) {
+            final int cmpDiff = ReadCoordinateComparator.compareCoordinates(previousRead, newRead, header);
+            if (cmpDiff == 1) {
+                throw new GATKException.ShouldNeverReachHereException(
+                        String.format("Reads must be coordinate sorted (earlier %s later %s)", previousRead, newRead));
+            }
+            if (cmpDiff != 0) {
+                reservoir.signalEndOfInput();
+                if ( reservoir.hasFinalizedItems() ) {
+                    finalizeReservoir();
```

Comment:
If the reservoir somehow had no finalized items after a positional change, wouldn't it get stuck in an "end of stream" state, since you're now calling signalEndOfInput() here? And wouldn't it then just explode on the subsequent reservoir.submit() call?

Perhaps we should just unconditionally call into `finalizeReservoir()` on every positional change, and get rid of the `signalEndOfInput()`/`hasFinalizedItems()` calls here (note that you're still calling `signalEndOfInput()` within `finalizeReservoir()` itself). That would ensure that the reservoir gets reset even if there were somehow no finalized items for a given position. It would also eliminate the need to call `signalEndOfInput()` twice as you're currently doing (you're calling it both here and in `finalizeReservoir()`)

Reply:
I don't think it should be possible to have a position change and not have any finalized items, so the `hasFinalizedItems`, and thats what required the redundant call to `signalEndOfInput`. So I think what you propose is both simpler and less redundant. Unless you can think of a counter-example, it might be worth adding an arg to `finalizeReservoir ` that trigger and assert that there are always finalized items.